packages/@blockprotocol/hook/src/react.ts

```diff
@@ -0,0 +1,246 @@
+import { RefObject, useEffect, useLayoutEffect, useRef, useState } from "react";
+
+import { HookBlockHandler, HookEmbedderHandler } from "./index.js";
+
+const useHookServiceConstructor = <
+  T extends typeof HookBlockHandler | typeof HookEmbedderHandler,
+>({
+  Handler,
+  constructorArgs,
+  ref,
+}: {
+  Handler: T;
+  constructorArgs?: Omit<ConstructorParameters<T>[0], "element">;
+  ref: RefObject<HTMLElement>;
+}) => {
+  const previousRef = useRef<HTMLElement | null>(null);
+
+  const [hookService, setHookService] = useState<
+    | (T extends typeof HookBlockHandler
+        ? HookBlockHandler
+        : HookEmbedderHandler)
+    | null
+  >(null);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps -- will not loop & we don't want to reconstruct on other args
+  useEffect(() => {
+    if (ref.current === previousRef.current) {
+      return;
+    }
+
+    if (previousRef.current) {
+      hookService?.destroy();
+    }
+
+    previousRef.current = ref.current;
+
+    if (ref.current) {
+      setHookService(
+        new Handler({
+          element: ref.current,
+          ...(constructorArgs as unknown as ConstructorParameters<T>), // @todo fix these casts
+        }) as T extends typeof HookBlockHandler // @todo fix these casts
+          ? HookBlockHandler
+          : HookEmbedderHandler,
+      );
+    }
+  });
+
+  return { hookService };
+};
+
+/**
+ * Create a HookBlockHandler instance, using a reference to an element in the
+ * block.
+ *
+ * The hookService will only be reconstructed if the element reference changes.
+ * Updates to any callbacks after first constructing should be made by
+ * calling hookService.on("messageName", callback);
+ */
+export const useHookBlockService = (
+  ref: RefObject<HTMLElement>,
+  constructorArgs?: Omit<
+    ConstructorParameters<typeof HookBlockHandler>[0],
+    "element"
+  >,
+): { hookService: HookBlockHandler | null } => {
+  return useHookServiceConstructor({
+    Handler: HookBlockHandler,
+    constructorArgs,
+    ref,
+  });
+};
+
+/**
+ * Create a HookBlockHandler instance, using a reference to an element in the
```

Comment:
This was originally my error in the graph service - 
```suggestion
 * Create a HookEmbedderHandler instance, using a reference to an element around the
```